src/Microsoft.AspNet.OData.Shared/Query/AutoSelectExpandHelper.cs

```diff
@@ -33,28 +34,22 @@ public static bool HasAutoSelectProperty(this IEdmModel edmModel, IEdmStructured
                 throw Error.ArgumentNull(nameof(edmModel));
             }
 
-            if (structuredType == null)
-            {
-                throw Error.ArgumentNull(nameof(structuredType));
-            }
-
-            List<IEdmStructuredType> structuredTypes = new List<IEdmStructuredType>();
-            structuredTypes.Add(structuredType);
-            structuredTypes.AddRange(edmModel.FindAllDerivedTypes(structuredType));
-
-            foreach (IEdmStructuredType edmStructuredType in structuredTypes)
+            if (structuredType != null)
             {
-                // for top type, let's retrieve its properties and the properties from base type of top type if has.
-                // for derived type, let's retrieve the declared properties.
-                IEnumerable<IEdmStructuralProperty> properties = edmStructuredType == structuredType
-                        ? edmStructuredType.StructuralProperties()
-                        : edmStructuredType.DeclaredStructuralProperties();
-
-                foreach (IEdmStructuralProperty subProperty in properties)
+                foreach (IEdmStructuredType edmStructuredType in new SelfAndDerivedEnumerator(structuredType, edmModel))
                 {
-                    if (IsAutoSelect(subProperty, property, edmStructuredType, edmModel))
+                    // for top type, let's retrieve its properties and the properties from base type of top type if has.
+                    // for derived type, let's retrieve the declared properties.
+                    IEnumerable<IEdmStructuralProperty> properties = edmStructuredType == structuredType
```

Comment:
> edmStructuredType == structuredType 

Wait, we are behaving different for the first element anyway? Why does it need to be part of the same sequence as the derived types? We can have an overload of `IsAutoSelect` that takes the properties sequence, and then this code becomes:
```
if (IsAutoSelect(structuredType.StructuralProperties))
{
  return true;
}

foreach (var edmStructuredType in edmModel.FindAllDerivedTypes(structuredType))
{
  if (IsAutoSelect(structuredType.StructuralProperties))
  {
    return true;
  }
}

return false;
```
Also, we seem to just be re-implementing `Any` here.

Reply:
This feedback seems to apply to other cases that are being updated in this PR as well

Reply:
Good callout - I didn't think too much about what the code was attempting to do, but on removing the list allocation. There appears to be further optimizations that we could do, but given the "PR complexity" analysis that is done I'd rather keep this one focused on the list change. We can follow up with more involved refactoring in another PR.

Reply:
I'll create an issue or something to track, I actually got lost in the code and didn't realize the behavior isn't actually changing here so I agree it's outside the scope of this change